### PR TITLE
Add JSDoc type annotations

### PR DESCRIPTION
**Pull request by y-hsgw**
I have added type annotations and fixed associated type errors.

### CODE REVIEW

File: lib/rules/button-has-type.js [hunk 1]
```diff
@@ -149,9 +150,11 @@ module.exports = {
         }
 
         const props = node.arguments[1].properties;
-        const typeProp = props.find((prop) => prop.key && prop.key.name === 'type');
+        const typeProp = props.find(
+          (prop) => prop.type === "Property" && prop.key && prop.key.type === "PrivateIdentifier" && prop.key.name === 'type'
+        );
 
-        if (!typeProp) {
+        if (typeProp.type !== "Property") {
```
Comment (ljharb) [inline lib/rules/button-has-type.js:157]: these changes aren't simple annotation changes. can they be extracted at least to a different commit, if not to a different PR?

File: lib/rules/checked-requires-onchange-or-readonly.js [hunk 2]
```diff
@@ -24,7 +24,7 @@ const defaultOptions = {
 };
 
 /**
- * @param {string[]} properties
+ * @param {Array} properties
```
Comment (ljharb) [inline lib/rules/checked-requires-onchange-or-readonly.js:27]: `Array` is less specific than `string[]`

File: lib/rules/checked-requires-onchange-or-readonly.js [hunk 3]
```diff
@@ -41,6 +41,7 @@ function extractTargetProps(properties, keyName) {
   );
 }
 
+/** @type { import('eslint').Rule.RuleModule } */
```
Comment (ljharb) [inline lib/rules/checked-requires-onchange-or-readonly.js:44]: curly braces should never have spaces directly inside them
```suggestion
/** @type {import('eslint').Rule.RuleModule} */
```

File: lib/rules/forbid-elements.js [hunk 4]
```diff
@@ -103,13 +104,11 @@ module.exports = {
...
-        const argType = argument.type;
-
-        if (argType === 'Identifier' && /^[A-Z_]/.test(argument.name)) {
+        if (argument.type === 'Identifier' && /^[A-Z_]/.test(argument.name)) {
           reportIfForbidden(argument.name, argument);
-        } else if (argType === 'Literal' && /^[a-z][^.]*$/.test(argument.value)) {
+        } else if (argument.type === 'Literal' && /^[a-z][^.]*$/.test(argument.value.toString())) {
           reportIfForbidden(argument.value, argument);
-        } else if (argType === 'MemberExpression') {
+        } else if (argument.type === 'MemberExpression') {
```
Comment (ljharb) [inline lib/rules/forbid-elements.js:111]: the original only has one observable lookup; this has 3. why is this an improvement?

also, `String(x)`, _never_ `x.toString()`

File: lib/rules/forbid-foreign-prop-types.js [hunk 5]
```diff
@@ -108,7 +109,7 @@ module.exports = {
             && !ast.isAssignmentLHS(node)
             && !isAllowedAssignment(node)
           )) || (
-            (node.property.type === 'Literal' || node.property.type === 'JSXText')
+            node.property.type === 'Literal'
             && node.property.value === 'propTypes'
```
Comment (ljharb) [inline lib/rules/forbid-foreign-prop-types.js:113]: this is meaningfully different; it will stop catching JSXText nodes.

**Review by ljharb — CHANGES_REQUESTED** [review]
this isn't just adding type annotations and a lot of the changes are potentially breaking.

File: lib/rules/button-has-type.js (hunk 1 above)
Comment (y-hsgw) [reply lib/rules/button-has-type.js:157]: I see. My apologies 🙇‍♂️. 
Would it be acceptable to create a PR that adds annotations only to rules where type changes are unnecessary?
Or would it be better to add annotations to all rules listed below, and if changes are needed, ignore the entire file (@ts-nocheck)?

Comment (ljharb) [reply lib/rules/button-has-type.js:157]: yes, that would be fine.

any non-type changes that are needed must be accompanied by tests - if the ostensible type error can't be reproduced, then the type system is wrong.

**Comment by y-hsgw** [comment]
I have addressed this in pull request #3732 and will proceed to close this one. I'll handle the areas requiring type changes separately.

**Comment by ljharb** [comment]
I would prefer to update this PR with those changes rather than abandoning it.

File: lib/rules/checked-requires-onchange-or-readonly.js (hunk 3 above)
Comment (y-hsgw) [reply lib/rules/checked-requires-onchange-or-readonly.js]: fixed in 9018532 .

File: lib/rules/forbid-elements.js (hunk 4 above)
Comment (y-hsgw) [reply lib/rules/forbid-elements.js]: > the original only has one observable lookup; this has 3. why is this an improvement?

The reason is because the following error message occurs during the execution of the type-check script.
`Property 'name' does not exist on type 'ClassExpression | ArrayExpression | ArrowFunctionExpression | AssignmentExpression | ... 23 more ... | SpreadElement'.  Property 'name' does not exist on type 'ClassExpression'.`

> also, String(x), never x.toString()

thanks! Fixed in 4ae24a2

File: lib/rules/checked-requires-onchange-or-readonly.js [hunk 6]
```diff
@@ -24,7 +24,7 @@ const defaultOptions = {
 };
 
 /**
- * @param {string[]} properties
+ * @param {Object[]} properties
```
Comment (ljharb) [inline lib/rules/checked-requires-onchange-or-readonly.js:27]:
```suggestion
 * @param {object[]} properties
```
let's use TS-jsdoc since that's the one that matters

File: lib/rules/forbid-foreign-prop-types.js [hunk 7]
```diff
@@ -121,7 +123,9 @@ module.exports = {
       },
 
       ObjectPattern(node) {
-        const propTypesNode = node.properties.find((property) => property.type === 'Property' && property.key.name === 'propTypes');
+        const propTypesNode = node.properties.find(
+          (property) => property.type === 'Property' && property.key.type === 'Identifier' && property.key.name === 'propTypes'
+        );
```
Comment (ljharb) [inline lib/rules/forbid-foreign-prop-types.js:128]: this change doesn't seem needed

File: lib/rules/forbid-prop-types.js [hunk 8]
```diff
@@ -1,3 +1,4 @@
+// @ts-nocheck
```
Comment (ljharb) [inline lib/rules/forbid-prop-types.js:1]: please remove this

File: lib/rules/jsx-closing-bracket-location.js [hunk 9]
```diff
@@ -195,7 +196,9 @@ module.exports = {
      */
     function getTokensLocations(node) {
       const sourceCode = context.getSourceCode();
+      // @ts-ignore
       const opening = sourceCode.getFirstToken(node).loc.start;
+      // @ts-ignore
```
Comment (ljharb) [inline lib/rules/jsx-closing-bracket-location.js:201]: please use ts-expect-error, with a comment, instead of ts-ignore.

File: lib/rules/jsx-curly-spacing.js [hunk 10]
```diff
@@ -1,3 +1,4 @@
+// @ts-nocheck
```
Comment (ljharb) [inline lib/rules/jsx-curly-spacing.js:1]:
```suggestion
```

File: lib/rules/jsx-equals-spacing.js [hunk 11]
```diff
@@ -61,7 +62,9 @@ module.exports = {
 
           const sourceCode = context.getSourceCode();
           const equalToken = sourceCode.getTokenAfter(attrNode.name);
+          // @ts-ignore
           const spacedBefore = sourceCode.isSpaceBetweenTokens(attrNode.name, equalToken);
+          // @ts-ignore
```
Comment (ljharb) [inline lib/rules/jsx-equals-spacing.js:67]: ts-expect-error

File: lib/rules/button-has-type.js [hunk 12]
```diff
@@ -149,9 +150,11 @@ module.exports = {
         }
 
         const props = node.arguments[1].properties;
-        const typeProp = props.find((prop) => prop.key && prop.key.name === 'type');
+        const typeProp = props.find(
+          (prop) => prop.type === 'Property' && prop.key && prop.key.type === 'PrivateIdentifier' && prop.key.name === 'type'
+        );
 
-        if (!typeProp) {
+        if (typeProp.type !== 'Property') {
```
Comment (ljharb) [inline lib/rules/button-has-type.js:157]: test?

File: lib/rules/jsx-fragments.js [hunk 13]
```diff
@@ -171,7 +172,7 @@ module.exports = {
       ImportDeclaration(node) {
         if (node.source && node.source.value === 'react') {
           node.specifiers.forEach((spec) => {
-            if (spec.imported && spec.imported.name === fragmentPragma) {
+            if (spec.type === 'ImportSpecifier' && spec.imported && spec.imported.name === fragmentPragma) {
```
Comment (ljharb) [inline lib/rules/jsx-fragments.js:175]: why is this needed? what are the possible `type` values of the objects contained an ImportDeclaration node's `specifiers` array - is it more than just `ImportSpecifier`?

File: lib/rules/jsx-indent-props.js [hunk 14]
```diff
@@ -140,6 +141,7 @@ module.exports = {
      * @return {Number} Indent
      */
     function getNodeIndent(node) {
+      // @ts-ignore
```
Comment (ljharb) [inline lib/rules/jsx-indent-props.js:144]: ts-expect-error

File: lib/rules/jsx-indent.js [hunk 15]
```diff
@@ -1,3 +1,4 @@
+// @ts-nocheck
```
Comment (ljharb) [inline lib/rules/jsx-indent.js:1]:
```suggestion
```

File: lib/rules/jsx-no-bind.js [hunk 16]
```diff
@@ -122,8 +123,8 @@ module.exports = {
       blockVariableNameSets[blockStart][violationType].add(variableName);
     }
 
-    function getBlockStatementAncestors(node) {
-      return context.getAncestors(node).filter(
+    function getBlockStatementAncestors() {
+      return context.getAncestors().filter(
```
Comment (ljharb) [inline lib/rules/jsx-no-bind.js:127]: i'm pretty sure this change will cause eslint 9 to break. can you confirm?

File: lib/rules/jsx-no-bind.js [hunk 17]
```diff
@@ -168,13 +169,15 @@ module.exports = {
...
           return;
         }
-        const blockAncestors = getBlockStatementAncestors(node);
+        const blockAncestors = getBlockStatementAncestors();
         const variableViolationType = getNodeViolationType(node.init);
 
         if (
           blockAncestors.length > 0
           && variableViolationType
+          && node.parent.type === 'VariableDeclaration'
```
Comment (ljharb) [inline lib/rules/jsx-no-bind.js:178]: we're already inside VariableDeclarator, this shouldn't be needed

File: lib/rules/jsx-sort-default-props.js [hunk 18]
```diff
@@ -172,6 +173,14 @@ module.exports = {
           return;
         }
 
+        if (
+          node.parent.type !== 'BinaryExpression'
+          && node.parent.type !== 'AssignmentExpression'
+          && node.parent.type !== 'LogicalExpression'
+        ) {
```
Comment (ljharb) [inline lib/rules/jsx-sort-default-props.js:180]: tests?

File: lib/rules/jsx-space-before-closing.js [hunk 19]
```diff
@@ -63,6 +64,21 @@ module.exports = {
...
+          && closingSlash.type !== 'Keyword'
+          && closingSlash.type !== 'Punctuator'
+          && closingSlash.type !== 'JSXIdentifier'
+          && closingSlash.type !== 'JSXText'
+          && closingSlash.type !== 'Numeric'
+          && closingSlash.type !== 'String'
+          && closingSlash.type !== 'RegularExpression'
+        ) {
+          return;
+        }
```
Comment (ljharb) [inline lib/rules/jsx-space-before-closing.js:80]: tests?

**Review by ljharb — CHANGES_REQUESTED** [review]
i only went like 20% through, but this should help you refine the entire thing

File: lib/rules/checked-requires-onchange-or-readonly.js (hunk 6 above)
Comment (y-hsgw) [reply lib/rules/checked-requires-onchange-or-readonly.js]: Thanks! Fixed in 844d483

File: lib/rules/forbid-foreign-prop-types.js (hunk 7 above)
Comment (y-hsgw) [reply lib/rules/forbid-foreign-prop-types.js]: Sorry🙇‍♂️  Fixed in https://github.com/jsx-eslint/eslint-plugin-react/commit/844d483855c56a6741cf56b0e3c9be2458d45e68 .

File: lib/rules/jsx-no-bind.js (hunk 17 above)
Comment (y-hsgw) [reply lib/rules/jsx-no-bind.js:178]: `Property 'kind' does not exist on type 'Node'.`
This error occurs.

File: lib/rules/jsx-fragments.js (hunk 13 above)
Comment (y-hsgw) [reply lib/rules/jsx-fragments.js:175]: > what are the possible type values of the objects contained an ImportDeclaration node's specifiers array - is it more than just ImportSpecifier?

`ImportSpecifier | ImportDefaultSpecifier | ImportNamespaceSpecifier` .

I added it because only ImportSpecifier has the imported property.

File: lib/rules/jsx-no-bind.js (hunk 16 above)
Comment (y-hsgw) [reply lib/rules/jsx-no-bind.js:127]: Sorry. I do not understand it well.
Should I add v9 to eslint in package.json as shown below and see if I get a compile error?
```
  "devDependencies": {
    "eslint": "^3 || ^4 || ^5 || ^6 || ^7 || ^8 || ^9",
    ...
  },
  "peerDependencies": {
    "eslint": "^3 || ^4 || ^5 || ^6 || ^7 || ^8 || ^9"
  },
```

Comment (ljharb) [reply lib/rules/jsx-no-bind.js:127]: No, adding eslint 9 is a hard task. What i mean is, what’s wrong with passing the node here?

File: lib/rules/jsx-no-bind.js (hunk 17 above)
Comment (ljharb) [reply lib/rules/jsx-no-bind.js:178]: if this is just for narrowing, we can use type casts to tell TS what kind of thing it is.

File: lib/rules/jsx-equals-spacing.js (hunk 11 above)
Comment (y-hsgw) [reply lib/rules/jsx-equals-spacing.js]: Thanks! Fixed in [eb44ba7](https://github.com/jsx-eslint/eslint-plugin-react/pull/3731/commits/eb44ba71aba5ed0d36518b3c658b7cab5e59c14f) .

File: lib/rules/jsx-indent-props.js (hunk 14 above)
Comment (y-hsgw) [reply lib/rules/jsx-indent-props.js]: Thanks! Fixed in [eb44ba7](https://github.com/jsx-eslint/eslint-plugin-react/pull/3731/commits/eb44ba71aba5ed0d36518b3c658b7cab5e59c14f) .

File: lib/rules/jsx-no-bind.js (hunk 16 above)
Comment (y-hsgw) [reply lib/rules/jsx-no-bind.js:127]: When looking at the type information for context.getAncestors, it seems that the argument is unnecessary.
``` ts
(method) Rule.RuleContext.getAncestors(): Node[]
```

File: lib/rules/jsx-no-bind.js (hunk 17 above)
Comment (y-hsgw) [reply lib/rules/jsx-no-bind.js:178]: In my understanding, since it's a JavaScript file, I don't think type assertions can be used.

File: lib/rules/jsx-closing-bracket-location.js (hunk 9 above)
Comment (y-hsgw) [reply lib/rules/jsx-closing-bracket-location.js]: Thanks! Fixed in [eb44ba7](https://github.com/jsx-eslint/eslint-plugin-react/pull/3731/commits/eb44ba71aba5ed0d36518b3c658b7cab5e59c14f) .

File: lib/rules/jsx-closing-bracket-location.js [hunk 20]
```diff
@@ -196,9 +196,11 @@ module.exports = {
      */
     function getTokensLocations(node) {
       const sourceCode = context.getSourceCode();
-      // @ts-ignore
+      // The types differ between ASTNode and ESTree.Node.
+      // @ts-expect-error
```
Comment (ljharb) [inline lib/rules/jsx-closing-bracket-location.js:200]:
```suggestion
      // @ts-expect-error The types differ between ASTNode and ESTree.Node.
```

Comment (y-hsgw) [reply lib/rules/jsx-closing-bracket-location.js]: Fixed in a5ac70f.

File: lib/rules/forbid-prop-types.js (hunk 8 above)
Comment (y-hsgw) [reply lib/rules/forbid-prop-types.js]: Fixed in[25e6afd](https://github.com/jsx-eslint/eslint-plugin-react/pull/3731/commits/25e6afd2a02e62af7d806fa6c2e12de46a97dcad).

File: lib/rules/jsx-indent.js (hunk 15 above)
Comment (y-hsgw) [reply lib/rules/jsx-indent.js]: Fixed in[25e6afd](https://github.com/jsx-eslint/eslint-plugin-react/pull/3731/commits/25e6afd2a02e62af7d806fa6c2e12de46a97dcad).

File: lib/rules/button-has-type.js [hunk 21]
```diff
@@ -149,9 +150,11 @@ module.exports = {
         }
 
         const props = node.arguments[1].properties;
```
Comment (ljharb) [inline lib/rules/button-has-type.js:152]: is the type of `node` properly narrowed to a CallExpression node here? if not, let's do that as an annotation on the visitor function rather than needing to add all these runtime changes.

Comment (y-hsgw) [reply lib/rules/button-has-type.js:155]: I attempted to make corrections in 040d77c, but is my understanding correct?

File: lib/rules/jsx-sort-default-props.js (hunk 18 above)
Comment (y-hsgw) [reply lib/rules/jsx-sort-default-props.js]: My apologies for the confusion. I mistakenly performed unnecessary type narrowing, so I removed it in commit d59ac20. Since the existing test cases seem to be unaffected, I did not add any additional tests.

File: lib/rules/jsx-space-before-closing.js (hunk 19 above)
Comment (y-hsgw) [reply lib/rules/jsx-space-before-closing.js]: My apologies for the confusion. I mistakenly performed unnecessary type narrowing, so I removed it in commit https://github.com/jsx-eslint/eslint-plugin-react/commit/d59ac204a929314b6c24956f7c3b31dd036a28fa. Since the existing test cases seem to be unaffected, I did not add any additional tests.

File: lib/rules/jsx-fragments.js (hunk 13 above)
Comment (ljharb) [reply lib/rules/jsx-fragments.js:175]: would `'imported' in spec` narrow properly?

File: lib/rules/jsx-no-bind.js (hunk 17 above)
Comment (ljharb) [reply lib/rules/jsx-no-bind.js:178]: they can, by wrapping the value in parens and preceding it with a `@type comment.

Comment (y-hsgw) [reply lib/rules/jsx-no-bind.js]: I didn't know that method! Thank you!  Fixed in e2c3497 .

File: lib/rules/jsx-fragments.js (hunk 13 above)
Comment (y-hsgw) [reply lib/rules/jsx-fragments.js:175]: Yes, it will be narrowed down. Alternatively, it might be better to do it as follows.

```js
 if (spec.type === 'ImportSpecifier' && spec.imported.name === fragmentPragma) { ...
```

File: lib/rules/button-has-type.js (hunk 21 above)
Comment (ljharb) [reply lib/rules/button-has-type.js:152]: That's an acceptable narrowing :-) thanks

**Review by ljharb — CHANGES_REQUESTED** [review]
Please rebase this PR, and extract any changes that aren't about whitespace/parens, or adding/editing comments, to separate commits - and then all the "type annotations" changes can be in a single commit i can review separately. Thanks!

**Comment by y-hsgw** [comment]
I have rebased and cleaned up the commits. Please check.

**Comment by ljharb** [comment]
I've gone ahead and updated this to remove all the breaking changes. The way to safely resolve TypeScript's inability to understand that an `x.foo` check is identical to `'foo' in x && x.foo` semantically is not to hardcode AST node types (that a custom parser might not match) but to add `'foo' in x` and let TS do the narrowing.

**Review by ljharb — APPROVED** [review]

File: lib/rules/no-is-mounted.js [hunk 22]
```diff
@@ -38,7 +39,10 @@ module.exports = {
         if (callee.type !== 'MemberExpression') {
           return;
         }
-        if (callee.object.type !== 'ThisExpression' || callee.property.name !== 'isMounted') {
+        if (
+          callee.object.type !== 'ThisExpression'
+          && (!('name' in callee.property) || callee.property.name !== 'isMounted')
+        ) {
```
Comment (voxpelli) [inline lib/rules/no-is-mounted.js:45]: This caused a regression, see https://github.com/jsx-eslint/eslint-plugin-react/issues/3819

It should always return when `callee.object.type` is not `'ThisExpression'` – no matter if `'name'` is missing in `callee.property` or not.

The functionality before this PR would be equivalent to this type checker compliant check:

```js
        if (
          callee.object.type !== 'ThisExpression' ||
          !('name' in callee.property) ||
          callee.property.name !== 'isMounted'
        ) {
```

Comment (voxpelli) [reply lib/rules/no-is-mounted.js:45]: PR with said fix: https://github.com/jsx-eslint/eslint-plugin-react/pull/3821
